skbio/tree/_nj.py

```diff
@@ -286,3 +287,332 @@ def _pair_members_to_new_node(dm, i, j, disallow_negative_branch_length):
+
+    Distances between nodes are calculated using a distance matrix.
+    """
+    nodelist1 = _tip_or_root(node1)
+    nodelist2 = _tip_or_root(node2)
+    df = dm.between(nodelist1, nodelist2)
+    return df["value"].mean()
+
+
+def _tip_or_root(node):
+    """Get name(s) of a node if it's a tip or root, otherwise its descending tips."""
+    if node.is_tip() or node.is_root():
+        return [node.name]
+    else:
+        return [x.name for x in node.tips()]
+
+
+def _average_distance_upper(node1, node2, dm):
+    """Returns the average distance between the leaves of two subtrees.
+
+    Used for subtrees which have a set of tips that are the complement
+    of the set of tips that are descendants from the node defining
+    the subtree.
+
+    Given an internal edge of a binary tree, exactly one adjacent edge
+    will connect to a node defining a subtree of this form.
+    """
+    nodelist1 = _tip_or_root(node1)
+    if node2.is_root():
+        nodelist2 = []
+    # Second subtree serves as the tree with a set of tips
+    # complementary to the set of tips that descend from the
+    # corresponding second node.
+    else:
+        root2 = node2.root()
+        nodelist2 = [root2.name]
+        nodelist2.extend(root2.subset() - node2.subset())
+    df = dm.between(nodelist1, nodelist2)
+    return df["value"].mean()
+
+
+def _subtree_count(subtree):
+    """Returns the number of leaves in a subtree.
+
+    Assumes the root as a leaf node.
+    """
+    if subtree.is_tip() or subtree.is_root():
+        return 1
+    else:
+        return subtree.count(tips=True)
+
+
+def _swap_length(a, b, c, d, i, j, k, m, adm):
+    """Returns the change in overall tree length after a given swap.
+
+    The count of leaves contained in each subtree are denoted 'a, b, c, d' while
+    each node defining the subtree has the index 'i, j, k, m', respectively.
+    """
+    lambda1 = (a * d + b * c) / ((a + b) * (c + d))
+    lambda2 = (a * d + b * c) / ((a + c) * (b + d))
+    return 0.5 * (
+        (lambda1 - 1) * (adm[i][k] + adm[j][m])
+        - (lambda2 - 1) * (adm[i][j] + adm[k][m])
+        - (lambda1 - lambda2) * (adm[i][m] + adm[j][k])
+    )
+
+
+def _swap_heap(tree, adm):
+    """Returns a maxheap ordered by the swap length for all possible swaps."""
+    heap = []
+    ordered = list(tree.postorder(include_self=False))
+    root = tree.root()
+    n_taxa = root.count(tips=True) + 1
+    # begin by finding nodes which are the child node of an internal edge
+    for node in ordered:
+        # ignore tips of the tree
+        if node.is_tip():
+            continue
+        # identify the parent and grandparent nodes
+        parent = node
+        a = parent.parent
+        # identify the index of each neighboring node
+        for index, node in enumerate(ordered):
+            if node == a:
+                i1 = index
+        for child in parent.children:
+            if child.is_tip():
+                continue
+            else:
```

Comment:
Since you already did `continue` in the previous line, the current `else` can be omitted.